typescript/rebalance_bot/rebalance.ts

```diff
@@ -0,0 +1,356 @@
+          quote,
+          trialAmount,
+          swapData,
+          percentage,
+        );
+
+        if (result.success) {
+          return result;
+        }
+
+        // Trial failed, notify and continue to next percentage
+        await this.notificationManager.notifyRebalanceFailure(
+          quote.direction,
+          percentage,
+          result.error || "Unknown error",
+          isLastTrial,
+        );
+      } catch (error) {
+        const errorMessage =
+          error instanceof Error ? error.message : String(error);
+        logger.error(`Trial ${(percentage * 100).toFixed(0)}% failed:`, error);
+
+        await this.notificationManager.notifyRebalanceFailure(
+          quote.direction,
+          percentage,
+          errorMessage,
+          isLastTrial,
+        );
+      }
+    }
+
+    return {
+      success: false,
+      direction: quote.direction,
+      percentage: 0,
+      inputAmount: 0n,
+      error: "All trials failed",
+    };
+  }
+
+  private calculateTrialAmount(inputAmount: bigint, percentage: number): bigint {
+    const scaledPercentage = BigInt(Math.round(percentage * Number(PERCENTAGE_PRECISION)));
+    return (inputAmount * scaledPercentage) / PERCENTAGE_PRECISION;
+  }
+
+  private async checkFlashLoanAvailability(
+    quote: RebalanceQuote,
+    trialAmount: bigint,
+  ): Promise<boolean> {
+    try {
+      const debtTokenAddress = this.config.tokens.debt.address;
+      let requiredFlashAmount: bigint;
+
+      if (quote.direction === 1) {
+        // Increase: estimate required flash in debt for trialInput collateral
+        const collateralInBase =
+          await this.contracts.core.convertFromTokenAmountToBaseCurrency(
+            trialAmount,
+            this.config.tokens.collateral.address,
+          );
+        requiredFlashAmount =
+          await this.contracts.core.convertFromBaseCurrencyToToken(
+            collateralInBase,
+            debtTokenAddress,
+          );
+      } else {
+        // Decrease: required flash equals trialInput debt
+        requiredFlashAmount = trialAmount;
+      }
+
+      const maxFlashLoan =
+        await this.contracts.flashLender.maxFlashLoan(debtTokenAddress);
+      const maxAllowed = maxFlashLoan / FLASH_LOAN_SAFETY_DIVISOR;
+
+      const available = requiredFlashAmount <= maxAllowed;
+      if (!available) {
+        logger.debug("Flash loan precheck failed:", {
+          required: requiredFlashAmount.toString(),
+          maxAllowed: maxAllowed.toString(),
+        });
+      }
+
+      return available;
+    } catch (error) {
+      logger.error("Flash loan availability check failed:", {
+        error: error instanceof Error ? error.message : String(error),
+        quote: {
+          direction: quote.direction,
+          inputAmount: quote.inputTokenAmount.toString()
+        },
+        trialAmount: trialAmount.toString()
+      });
+      return false; // Fail safe: skip trial if precheck fails
+    }
+  }
+
```

Comment:
Good implementation of retry logic\! The exponential backoff and detection of retryable errors is well done. Consider also adding `'ECONNREFUSED'` to the list of retryable error messages for when the RPC endpoint is temporarily unavailable.